src/main/cpp/src/shuffle_split.cu

```diff
@@ -770,212 +539,247 @@ struct dst_offset_output_iterator {
+ *
+ * @param out_buffer The output buffer for the entire split operation
+ * @param out_buffer_offsets Per-partition offsets into the output buffer
+ * @param num_partitions The number of partitions
+ * @param columns_per_partition The number of flattened columns per partition
+ * @param split_indices Per-partition row split indices
+ * @param flattened_col_has_validity Per-column bool on whether each column contains a validity
+ * vector
+ * @param partition_size_info Per-partition size information for each buffer type
  *
- * Note: columns types which themselves inherently have no data (strings, lists,
- * structs) return 0.
  */
-struct size_of_helper {
-  template <typename T>
-  constexpr std::enable_if_t<not is_fixed_width<T>(), int> __device__ operator()() const
-  {
-    return 0;
+__global__ void pack_per_partition_metadata_kernel(uint8_t* out_buffer,
+                                                   size_t const* out_buffer_offsets,
+                                                   size_t num_partitions,
+                                                   size_t columns_per_partition,
+                                                   size_type const* split_indices,
+                                                   int8_t const* flattened_col_has_validity,
+                                                   partition_size_info const* partition_sizes)
+{
+  constexpr uint32_t magic = 0x4b554430;
+
+  int const tid = threadIdx.x + (blockIdx.x * blockDim.x);
+  auto const threads_per_partition =
+    cudf::util::round_up_safe(columns_per_partition, static_cast<size_t>(cudf::detail::warp_size));
+  auto const partition_index = tid / threads_per_partition;
+  if (partition_index >= num_partitions) { return; }
+  auto const col_index = tid % threads_per_partition;
+
+  // start of the metadata buffer for this partition
+  uint8_t* buf_start        = out_buffer + out_buffer_offsets[partition_index];
+  partition_header* pheader = reinterpret_cast<partition_header*>(buf_start);
+
+  // first thread in each partition stores constant stuff
+  if (col_index == 0) {
+    pheader->magic_number = cudf::hashing::detail::swap_endian(magic);
+
+    pheader->row_index =
+      cudf::hashing::detail::swap_endian(static_cast<uint32_t>(split_indices[partition_index]));
+
+    // it is possible to get in here with no columns -or- no rows.
+    auto const partition_num_rows =
+      col_index < columns_per_partition
+        ? split_indices[partition_index + 1] - split_indices[partition_index]
+        : 0;
+    pheader->num_rows =
+      cudf::hashing::detail::swap_endian(static_cast<uint32_t>(partition_num_rows));
+
+    auto const& psize = partition_sizes[partition_index];
+    pheader->validity_size =
+      cudf::hashing::detail::swap_endian(static_cast<uint32_t>(psize.validity_size));
+    pheader->offset_size =
+      cudf::hashing::detail::swap_endian(static_cast<uint32_t>(psize.offset_size));
```

Comment:
The validity and offset sizes in the CPU version include the padding at the end.

Reply:
@nvdbaranec Just as a side note I was able to update the CPU code to match the padding that you are doing on the GPU code. So the only thing that is left to really make these cross compatible is the changes for the lengths.

Reply:
Added the padding to the sizes and changed `data_size` to `total_size`, representing (validity+offsets+data).